kamon-annotation/src/main/scala/kamon/annotation/instrumentation/BaseAnnotationInstrumentation.scala

```diff
@@ -0,0 +1,180 @@
+      val category = eval(segment.category())
+      val library = eval(segment.library())
+      val metadata = tags(segment.tags())
+      segments.put(methodName(method), SegmentInfo(name, category, library, metadata))
+    }
+  }
+
+  @inline final def processTrace(traces: TrieMap[String, TraceContext], pjp: ProceedingJoinPoint): AnyRef = {
+    val name = methodName(pjp.getStaticPart.getSignature.asInstanceOf[MethodSignature].getMethod)
+    traces.get(name).map { traceContext ⇒
+      Tracer.withContext(traceContext) {
+        val result = pjp.proceed()
+        Tracer.currentContext.finish()
+        result
+      }
+    } getOrElse pjp.proceed()
+  }
+
+  @inline final def processSegment(segments: TrieMap[String, SegmentInfo], pjp: ProceedingJoinPoint): AnyRef = {
+    val name = methodName(pjp.getStaticPart.getSignature.asInstanceOf[MethodSignature].getMethod)
+    segments.get(name).map { segment ⇒
+      Tracer.currentContext.collect { ctx ⇒
+        val currentSegment = ctx.startSegment(segment.name, segment.category, segment.library)
+        segment.tags.foreach { case (key, value) ⇒ currentSegment.addMetadata(key, value) }
+        val result = pjp.proceed()
+        currentSegment.finish()
+        result
+      } getOrElse pjp.proceed()
+    } getOrElse pjp.proceed()
+  }
+
+  @inline final def processTime(histograms: TrieMap[String, kamon.metric.instrument.Histogram], pjp: ProceedingJoinPoint): AnyRef = {
+    val name = methodName(pjp.getStaticPart.getSignature.asInstanceOf[MethodSignature].getMethod)
+    histograms.get(name).map(Latency.measure(_)(pjp.proceed)).getOrElse(pjp.proceed())
+  }
+
+  @inline final def processHistogram(histograms: TrieMap[String, kamon.metric.instrument.Histogram], result: AnyRef, jps: JoinPoint.StaticPart): Unit = {
+    val name = methodName(jps.getSignature.asInstanceOf[MethodSignature].getMethod)
+    val histogram = histograms.get(name)
+    histogram.map(_.record(result.asInstanceOf[Number].longValue()))
+  }
+
+  final def processCount(counters: TrieMap[String, kamon.metric.instrument.Counter], pjp: ProceedingJoinPoint): AnyRef = {
+    val name = methodName(pjp.getStaticPart.getSignature.asInstanceOf[MethodSignature].getMethod)
+    val counter = counters.get(name)
+    try pjp.proceed() finally counter.map(_.increment())
+  }
+
+  final def processMinMax(minMaxCounters: TrieMap[String, kamon.metric.instrument.MinMaxCounter], pjp: ProceedingJoinPoint): AnyRef = {
+    val name = methodName(pjp.getStaticPart.getSignature.asInstanceOf[MethodSignature].getMethod)
+    val minMaxCounter = minMaxCounters.get(name)
+    minMaxCounter.map(_.increment())
+    try pjp.proceed() finally minMaxCounter.map(_.decrement())
+  }
+
+  private[this] def methodName(method: Method): String = method.toString.replace(" ", "-").toLowerCase
+}
+
+@Aspect
+class ClassToProfiledMixin {
+  @DeclareMixin("(@kamon.annotation.Metrics *)")
+  def mixinClassToProfiled: Profiled = new Profiled {}
+}
+
+trait Profiled {
```

Comment:
What about `AnnotationInstruments`? Also, as pointed out above, the traces member should not map to a TraceContext. I guess the reason why this works is that we are not guarding against finishing the same TraceContext multiple times, do you think we should?

Finally, taking processHistogram as an example: it is calling histograms.get(name).map(...), but since we are registering the required instruments upon the creation of the annotated class instance, we could just do histograms.apply(..) and avoid mapping over the Option[Histogram]. If the required instrument is not available in the map it means we have a bug to fix or that the class wasn't annotated with `@EnableKamonAnnotations`, both cases would need manual intervention to get fixed.